docs: Add rules meta info to rule pages

nzakas opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[x] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

This PR makes several changes to the rules doc pages:

1. It inserts the description from the JavaScript rule source file under the header.
2. It shows panels for recommended, fixable, and suggestions at the top of the pages.
3. It updates Makefile.js to include functionality to generate a rules meta data JSON file for use in the docs site. This will be done automatically during site generation of the current eslint.org so it will be updated with each release.

#### Is there anything you'd like reviewers to focus on?

Don't worry too much about the styling right now...I still need to fix that. 

In the new site, the description tends to be duplicated because each rule's first line of content is the rule description. We can fix that once we move over to the new site.

Otherwise, does everything appear in the correct spot?

<!-- markdownlint-disable-file MD004 -->

btmills commented on docs/src/_includes/layouts/doc.html:
```diff
@@ -35,9 +37,20 @@ <h2 id="further-reading">Further Reading</h2>
+                <p>{{ rule_meta.docs.description | markdown | safe }}</p>
+                {{ ruleCategories({
+                recommended: rule_meta.docs.recommended,
+                fixable: rule_meta.fixable,
```
The fixable blurb isn't showing on e.g. [`arrow-body-style`](https://deploy-preview-15902--docs-eslint.netlify.app/rules/arrow-body-style/), perhaps because the `ruleCategories` macro [expects this to be a boolean](https://github.com/eslint/eslint/blob/901ce0f1e32ea1e9e10ce4d8b37c0d750007a3c5/docs/src/_includes/components/rule-categories.macro.html#L12), but `rule_meta.fixable` will be either `"code"` or `"whitespace"`.

nzakas replied:
Oops, good call.

mdjermanovic commented on Makefile.js:
```diff
@@ -181,10 +181,14 @@ function generateFormatterExamples(formatterInfo, prereleaseVersion) {
 function generateRuleIndexPage() {
-    const outputFile = "../website/_data/rules.yml",
+    const legacyWebsiteOutputFile = "../website/_data/rules.yml",
+        docsSiteOutputFile = "docs/src/_data/rules.json",
+        docsSiteMetaOutputFile = "docs/src/_data/rules_meta.json",
```
I believe the changes in these two files will not be committed during the release. Our release process commits only [changelog](https://github.com/eslint/eslint-release/blob/198c7e700e501f2762f41396635b74fdb7d04e06/lib/release-ops.js#L381-L382) and [version](https://github.com/eslint/eslint-release/blob/198c7e700e501f2762f41396635b74fdb7d04e06/lib/release-ops.js#L385) (`npm version` creates a version commit and tag).

nzakas replied:
Good call. Maybe we should update eslint-release to run an `npm build` command and then commit the result?

mdjermanovic replied:
> Good call. Maybe we should update eslint-release to run an `npm build` command and then commit the result?

Maybe we could add this as an option in [`generateRelease()`](https://github.com/eslint/eslint-release/blob/198c7e700e501f2762f41396635b74fdb7d04e06/lib/release-ops.js#L365)? Some projects don't have a `build` script (for example, [`eslint-plugin-markdown`](https://github.com/eslint/eslint-plugin-markdown/blob/87c2b536fd80b15e134766d92b90048ae45cbe1f/package.json#L22-L32)), and some project have a `build` script that should not be run for this purpose (for example, [`espree`](https://github.com/eslint/espree/blob/c98a7e0be7e21f28a00868660d66bda9d323f1d3/package.json#L70)).

mdjermanovic replied:
Alternatively, we could update and commit these files before running [`ReleaseOps.generateRelease()`](https://github.com/eslint/eslint/blob/ce035e5fac632ba8d4f1860f92465f22d6b44d42/Makefile.js#L267). A downside is that the commit would be included in the changelog.

Or add a `preversion` or `version` script ([npm version](https://docs.npmjs.com/cli/v8/commands/npm-version)). A downside is that this would bundle the changes into the version commit, I think.

nzakas replied:
I’ll dig into this more.

mdjermanovic replied:
> we can add the new files and amend the commit after we generate the release.

Version tag will still point to the original version commit, which won't be in the history of the `main` branch anymore. I think that's not a common practice.

nzakas replied:
Hmm good point. I suppose the next logical step is just to update the tag, too.

nzakas commented:
Updated the git tag after modifying the commit and rebased on top of main.

Note: I noticed a UI bug related to the displaying of fixable/suggestions/recommended on rules. I've opened https://github.com/eslint/eslint/issues/15977 to track separately.

nzakas commented:
@btmills can you take a look at this to clear your review?

btmills reviewed (approved):
I commented on #15977 as well, but I'm seeing all three categories displaying correctly now!

nzakas commented on Makefile.js:
```diff
@@ -271,6 +292,10 @@ function generateRelease() {
+
+    echo("Updating commit with docs data");
+    exec("git add docs/src/_data && git commit --amend --no-edit");
+    exec(`git tag -f v${releaseInfo.version}`);
```
I’ll have to check if the annotation remains after this or if it’s lost.

nzakas replied:
It looks like everything is lost, so I updated to create a new annotated tag with the version as a message (to duplicate what `npm version` does).

mdjermanovic reviewed (approved):
LGTM, thanks!
